features/FxA-54-connect-another-device-phase-3/README.md

```diff
@@ -0,0 +1,367 @@
+   it is opened instead
+   and the installation step is skipped.
+
+6. User installs Firefox on their device.
+
+7. User opens Firefox and is taken to the FxA sign-in screen,
+   with their email address pre-filled.
+
+8. User signs in and their device is connected to Sync.
+
+## Implementation
+
+### Content server
+
+* A `features` parameter will be submitted to `POST /sms`,
+  containing the value `[ "signinCodes" ]`.
+
+* An endpoint will be added for redirecting users through Adjust
+  to the appropriate app store, `GET /m/:signinCode`.
+
+* The sign-in screen will read the `signin` query parameter
+  and submit the value to `POST /signinCodes/consume`.
+  The email address in the sign-in form will be prefilled
+  using the response from that request.
+  If the request fails for any reason,
+  the user will not see any errors.
+  Instead they can just continue signing in
+  by typing in their email address themselves.
+
+### Auth server
+
+* If the `signinCodes` feature is specified to `POST /sms`,
+  the SMS message will include a different URL
+  using the content server's `/m/:signinCode` redirection endpoint
+  and a freshly-created signinCode.
+
+* A `POST /signinCodes/consume` endpoint will be added,
+  which takes a signinCode and returns an email address.
+  If the signinCode is invalid, the request will fail.
+
+### Auth database
+
+* A signinCodes table will be added,
+  containing the columns `signinCode`, `uid` and `createdAt`.
+
+* A `PUT /signinCodes/:signinCode` endpoint will be added,
+  which writes a signinCode to the database.
+  A signinCode will be valid for two days
+  from the time at which it is created.
+
+* A `POST /signinCodes/:signinCode/consume` endpoint will be added,
+  which removes valid signinCodes from the database
+  and returns the appropriate email address.
+
+* As part of the existing token-pruning loop,
+  signinCodes that are older than three months
+  will be expunged.
+
+### Mobile clients
+
+* Support will be added for deep linking via Adjust.
+  This includes all of the special magic
+  for redirecting to the app store
+  and propagating signinCodes to the FxA sign-in screen.
+  Both mobile clients will use the following URL scheme:
+
+  ```
+  firefox://fxa-signin?signin=<signinCode>&entrypoint=<entryPoint>&utm_*=<value>
+  ```
+
+### Mock-ups
+
+When the `/signin` page is opened with a valid signinCode, the email address
+will not be editable.
+
+<img src="./images/signin_valid_code_ios.png" alt="signin opened with valid signinCode" width="250px"/>
+
+### Test plan/acceptance criteria
+
+Reference images are below each OS section.
+
+signinCodes are disabled by default and can be enabled by opening the signup
+verification link with the `signinCodes=true` query parameter. Note, sending
+SMS is only available during the signup flow, signing into an existing account
+will not trigger the flow.
+
+#### Android
+
+##### Testing without support for deep links in Firefox for Android
```

Comment:
Fx Beta channels support deep linking, and once https://github.com/mozilla/fxa-content-server/pull/5221 makes its way to prod, we'll be able to remove the "Testing without support for deep links" sections for both Firefox for Android and iOS. Instead of "testing without support", we can update the instructions to open the SMS link with the `channel=(beta|nightly)` to install different versions of Firefox.